src/machine/machine_stm32_i2c.go

```diff
@@ -0,0 +1,481 @@
+	} else {
+		return i2c.Bus.SR2.HasBits(flag & mask)
+	}
+}
+
+func (i2c I2C) clearFlag(flag uint32) {
+	const mask = 0x0000FFFF
+	i2c.Bus.SR1.Set(^(flag & mask))
+}
+
+// clearFlagADDR reads both status registers to clear any pending ADDR flags.
+func (i2c I2C) clearFlagADDR() {
+	i2c.Bus.SR1.Get()
+	i2c.Bus.SR2.Get()
+}
+
+func (i2c I2C) waitForFlag(flag uint32, set bool) bool {
+	const tryMax = 10000
+	hasFlag := false
+	for i := 0; !hasFlag && i < tryMax; i++ {
+		hasFlag = i2c.hasFlag(flag) == set
+	}
+	return hasFlag
+}
+
+func (i2c I2C) waitForFlagOrError(flag uint32, set bool) bool {
+	const tryMax = 10000
+	hasFlag := false
+	for i := 0; !hasFlag && i < tryMax; i++ {
+		if hasFlag = i2c.hasFlag(flag) == set; !hasFlag {
+			// check for ACK failure
+			if i2c.hasFlag(flagAF) {
+				// generate stop condition
+				i2c.Bus.CR1.SetBits(stm32.I2C_CR1_STOP)
+				// clear pending flags
+				i2c.clearFlag(flagAF)
+				return false
+			} else if i2c.hasFlag(flagSTOPF) {
+				// clear stop flag
+				i2c.clearFlag(flagSTOPF)
+				return false
+			}
+		}
+	}
+	return hasFlag
+}
+
+type transferOption uint32
+
+const (
+	frameFirst        = 0x00000001
+	frameFirstAndNext = 0x00000002
+	frameNext         = 0x00000004
+	frameFirstAndLast = 0x00000008
+	frameLastNoStop   = 0x00000010
+	frameLast         = 0x00000020
+	frameNoOption     = 0xFFFF0000
+)
+
+// addressable represents a type that can provide fully-formatted I2C slave
+// addresses for both read operations and write operations.
+type addressable interface {
+	toRead() uint32
+	toWrite() uint32
+	bitSize() uint8
+}
+
+// address7Bit and address10Bit stores the unshifted original I2C slave address
+// in an unsigned integral data type and implements the addressable interface
+// to reformat addresses as required for read/write operations.
+//   TODO:
+//    add 10-bit address support
+type (
+	address7Bit uint8
+	//address10Bit uint16
+)
+
+func (sa address7Bit) toRead() uint32 {
+	return uint32(((uint8(sa) << 1) | uint8(stm32.I2C_OAR1_ADD0)) & 0xFF)
+}
+func (sa address7Bit) toWrite() uint32 {
+	return uint32(((uint8(sa) << 1) & ^(uint8(stm32.I2C_OAR1_ADD0))) & 0xFF)
+}
+func (sa address7Bit) bitSize() uint8 { return 7 } // 7-bit addresses
+
+//func (sa address10Bit) toRead() uint32  {}
+//func (sa address10Bit) toWrite() uint32 {}
+//func (sa address10Bit) bitSize() uint8  { return 10 } // 10-bit addresses
+
+func readAddress7Bit(addr uint8) uint32  { return address7Bit(addr).toRead() }
+func writeAddress7Bit(addr uint8) uint32 { return address7Bit(addr).toWrite() }
+
+//func readAddress10Bit(addr uint16) uint32  { return address10Bit(addr).toRead() }
+//func writeAddress10Bit(addr uint16) uint32 { return address10Bit(addr).toWrite() }
+
+// I2C fast mode (Fm) duty cycle
+const (
+	DutyCycle2    = 0
+	DutyCycle16x9 = 1
+)
+
+// I2CConfig is used to store config info for I2C.
+type I2CConfig struct {
+	Frequency uint32
+	SCL       Pin
+	SDA       Pin
+	DutyCycle uint8
```

Comment:
What does `DutyCycle` mean? How is it different from `Frequency`?

Reply:
I was confused by this one too. It seems to represent some constant factor that affects the I2C Fast Mode frequency calculation only (Standard Mode is unaffected). Unfortunately, I'm not sure if it is something we can determine automatically (based on requested frequency) or if it's an independent setting the user needs access to in order to configure the bus clock. Therefore, I left it available for the user.

You can see how I'm using it to calculate frequency ([machine_stm32f405.go:109-115](https://github.com/ardnew/tinygo/blob/feature/feather-stm32f405-i2c/src/machine/machine_stm32f405.go#L109-L115)), which, again, comes from a combination of the [STM32 programming manual](https://www.st.com/resource/en/reference_manual/dm00031020-stm32f405415-stm32f407417-stm32f427437-and-stm32f429439-advanced-armbased-32bit-mcus-stmicroelectronics.pdf) and the STM32CubeIDE auto-generated HAL and LL I2C drivers (in C).

For reference, I've copied the register definition below, which is basically the only location this `DutyCycle` (`DUTY`) is discussed, but it also includes an example calculation. Maybe you or someone else can infer how to best handle this item?

```
27.6.8   I2C Clock control register (I2C_CCR)

            Address offset:  0x1C
            Reset value:     0x0000

    Note:   fPCLK1 must be at least 2 MHz to achieve Sm mode I²C frequencies. It must be at least 4
            MHz to achieve Fm mode I²C frequencies. It must be a multiple of 10MHz to reach the
            400 kHz maximum I²C Fm mode clock.

            The CCR register must be configured only when the I2C is disabled (PE = 0).


  +------+------+------+------+------+------+------+------+------+------+------+------+------+------+------+------+
  |  15  |  14  |  13  |  12  |  11  |  10  |  9   |  8   |  7   |  6   |  5   |  4   |  3   |  2   |  1   |  0   |
  +------+------+------+------+------+------+------+------+------+------+------+------+------+------+------+------+
  | F/S  | DUTY |             |                                     CCR[11:0]                                     |
  +------+------+  Reserved   +------+------+------+------+------+------+------+------+------+------+------+------+
  |  rw  |  rw  |             |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |  rw  |
  +------+------+------+------+------+------+------+------+------+------+------+------+------+------+------+------+


            Bit 15 F/S: I2C master mode selection

                    0: Sm mode I2C
                    1: Fm mode I2C


            Bit 14 DUTY: Fm mode duty cycle

                    0: Fm mode Tlow/Thigh = 2
                    1: Fm mode Tlow/Thigh = 16/9 (see CCR)


            Bits 13:12 Reserved, must be kept at reset value


            Bits 11:0 CCR[11:0]: Clock control register in Fm/Sm mode (Master mode)
                                 Controls the SCL clock in master mode.

                    Sm mode or SMBus:
                        Thigh = CCR * TPCLK1
                        Tlow = CCR * TPCLK1
                    Fm mode:
                        If DUTY = 0:
                            Thigh = CCR * TPCLK1
                            Tlow = 2 * CCR * TPCLK1
                        If DUTY = 1: (to reach 400 kHz)
                            Thigh = 9 * CCR * TPCLK1
                            Tlow = 16 * CCR * TPCLK1

                    For instance: in Sm mode, to generate a 100 kHz SCL frequency:
                        If FREQR = 08, TPCLK1 = 125 ns so CCR must be programmed with 0x28 
                            (0x28 <=> 40d x 125 ns = 5000 ns.)


            Note:   The minimum allowed value is 0x04, except in FAST DUTY mode where the minimum
                    allowed value is 0x01.

                    Thigh = tr(SCL) + tw(SCLH). See device datasheet for the definitions of parameters.
                    Tlow = tf(SCL) + tw(SCLL). See device datasheet for the definitions of parameters.

                    I2C communication speed, fSCL ~ 1/(thigh + tlow). The real frequency may differ due to
                    the analog noise filter input delay.

                    The CCR register must be configured only when the I2C is disabled (PE = 0).

```

Also, if you are at all interested in the STM32 driver code a lot of these PRs are derived from, I've copied them here:

https://github.com/ardnew/ardnew/tree/master/share/tinygo/sago35

ST provides 2 types of drivers for their peripherals: a HAL driver and a LL (low-level) driver. Both are included as separate packages in that repo. The relevant files would be, for example, `feather-stm32f405/Drivers/STM32F4xx_HAL_Driver/stm32f4xx_hal_i2c.c`